systems/ps5.ixx

```diff
@@ -0,0 +1,151 @@
+
+        it = param_json.find("masterDataId");
+        if(it != param_json.end())
+            os << std::format("  serial: {}", it->second) << std::endl;
+    }
+
+private:
+    std::map<std::string, std::string> loadJSON(std::shared_ptr<iso9660::Entry> root_directory, std::string json_file) const
+    {
+        std::map<std::string, std::string> json;
+
+        auto json_entry = root_directory->subEntry(json_file);
+        if(json_entry)
+        {
+            auto data = json_entry->read();
+            if(data.size() <= 0x800)
+                return json;
+
+            data.erase(data.begin(), data.begin() + 0x800);
+
+            // Parse JSON into key/value pairs
+            size_t cur = 0;
+            while(cur < data.size())
+            {
+                // Find start of key
+                while(cur < data.size())
+                {
+                    if(data[cur] == '"')
+                        break;
+                    ++cur;
+                }
+                if(cur >= data.size())
+                    break;
+                ++cur;
+                size_t keyStart = cur;
+
+                // Find end of key
+                while(cur < data.size())
+                {
+                    if(data[cur] == '"')
+                        break;
+                    ++cur;
+                }
+                if(cur >= data.size())
+                    break;
+                size_t keyEnd = cur;
+                ++cur;
+
+                // Find start of value
+                while(cur < data.size())
+                {
+                    if(data[cur] == ':')
+                        break;
+                    ++cur;
+                }
+                if(cur >= data.size())
+                    break;
+                ++cur;
+                size_t valueStart = cur;
+
+                // Find end of value
+                while(cur < data.size())
+                {
+                    if(data[cur] == '[' || data[cur] == ',' || data[cur] == '}')
+                        break;
+                    ++cur;
+                }
+                if(cur >= data.size())
+                    break;
+                size_t valueEnd = cur;
+                ++cur;
+
+                // Don't parse arrays, treat JSON as flat
+                if(data[cur - 1] == '[')
+                    continue;
+
+                // Assign key and value
+                std::string key(data.begin() + keyStart, data.begin() + keyEnd);
+                std::string value(data.begin() + valueStart, data.begin() + valueEnd);
+                erase_all_inplace(key, '\0');
+                erase_all_inplace(key, '\r');
+                erase_all_inplace(key, '\n');
+                erase_all_inplace(value, '\0');
+                erase_all_inplace(value, '\r');
+                erase_all_inplace(value, '\n');
+                trim_inplace(key);
+                trim_inplace(value);
```

Comment:
Removing `\0 \r \n` seem an overkill to me (unless serial and version that we use have those characters).
There is also an easy way to get rid of trailing \0 while being safe if string doesn't end with \0, example:
https://github.com/superg/redumper/blob/main/systems/ps3.ixx#L183

Reply:
Removing new lines is needed in order to properly trim the values due to the naive way I am parsing JSON. As values are not always enclosed in quotations, sometimes it is needed to keep looking until the end } which occurs after a newline. Removing nulls is just for safety.

Reply:
Ah, got it, so key/value can be surrounded by the new lines that get skipped by parser. Keep it then.